spark/src/main/scala/org/apache/spark/sql/delta/DeltaLog.scala

```diff
@@ -439,16 +439,27 @@ class DeltaLog private(
   def ensureLogDirectoryExist(): Unit = {
     val fs = logPath.getFileSystem(newDeltaHadoopConf())
     def createDirIfNotExists(path: Path): Unit = {
-      var success = false
       try {
-        success = fs.exists(path) || fs.mkdirs(path)
+        // Optimistically attempt to create the directory first without checking its existence.
+        // This is efficient because we're assuming it's more likely that the directory doesn't
+        // exist and it saves an filesystem existence check in that case.
+        if (!fs.mkdirs(path) && !fs.exists(path)) {
+          throw DeltaErrors.cannotCreateLogPathException(logPath.toString)
+        }
       } catch {
+        // Cloud Hadoop File Systems implementations are expected to throw
+        // org.apache.hadoop.fs.FileAlreadyExistsException if the file already exists.
+        // e.g. S3AFileSystem.
+        case _: org.apache.hadoop.fs.FileAlreadyExistsException => // Ignore this exception
+        // Unix-based File Systems might throw java.nio.file.FileAlreadyExistsException
+        // if the file already exists.
+        case _: java.nio.file.FileAlreadyExistsException => // Ignore this exception
```

Comment:
These two catches are unsafe -- the docs for [Files.createDirectories](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#createDirectories-java.nio.file.Path-java.nio.file.attribute.FileAttribute...-) state that 
> an exception is not thrown if the directory could not be created because it already exists. 

... except for one special case that we definitely don't want to ignore:
>  `FileAlreadyExistsException` - if dir exists but is not a directory (optional specific exception)

Thus, we can only ignore the exception if we prove the existing path is a directory.

Reply:
I think the safest approach is to adjust the existing code to do something like this:
```scala
var success = try {
  // NOTE: Return value of false should mean the directory already existed (not an error) but
  // we will verify below because we're paranoid about buggy FileSystem implementations.
  fs.mkdirs(...)
} catch {
  // NOTE: Only needed because buggy Hadoop FileSystem.mkdir wrongly throws on existing dir.
  case _: IOException => false
}
if (!success) {
  success = try {
    fs.getFileStatus(...).isDir
  } catch {
    NonFatal(_) => false
  }
}
if (!success) {
  throw DeltaErrors.cannotCreateLogPathException(...)
}
```

Reply:
Added this + some handling for IOExceptions (which we already had)